### PR TITLE
fix(config): `enoding.only_fields` should properly handle parent keys

### LINKED ISSUE #2407: encoding.only_fields breaks certain kinds of serialization
### Vector Version

```
vector 0.9.0 (v0.9.0 x86_64-unknown-linux-musl 2020-04-20)
```

### Vector config file

```yaml
[sources.stdin]
type = "stdin"

[transforms.with_metadata]
type = "add_fields"
inputs = ["stdin"]
fields.tags = ["my-tag"]
fields.parent.child = ["my-child"]

[sinks.output]
type = "console"
inputs = ["with_metadata"]
encoding.codec = "json"
encoding.only_fields = ["message", "parent", "tags"]
```

### Debug output

```
Apr 22 14:34:39.882  INFO vector: Log level "trace" is enabled.
Apr 22 14:34:39.883  INFO vector: Loading configs. path=["faulty.yml"]
Apr 22 14:34:39.883 TRACE vector: Parsing config. path="faulty.yml"
Apr 22 14:34:39.896  INFO vector: Vector is starting. version="0.9.0" git_version="v0.9.0" released="Mon, 20 Apr 2020 21:15:23 +0000" arch="x86_64"
Apr 22 14:34:39.897  INFO vector::topology: Running healthchecks.
Apr 22 14:34:39.897  INFO vector::topology: Starting source "stdin"
Apr 22 14:34:39.897  INFO vector::topology: Starting transform "with_metadata"
Apr 22 14:34:39.897  INFO vector::topology: Starting sink "output"
Apr 22 14:34:39.899  INFO vector::topology::builder: Healthcheck: Passed.
Apr 22 14:34:39.899  INFO source{name=stdin type=stdin}: vector::sources::stdin: Capturing STDIN
Apr 22 14:34:39.899 TRACE sink{name=output type=console}: vector::sinks::util::sink: Acking events. acking_num=0
Apr 22 14:34:39.903  INFO source{name=stdin type=stdin}: vector::sources::stdin: finished sending
Apr 22 14:34:39.904 TRACE sink{name=output type=console}: vector::sinks::util::sink: sending item.
Apr 22 14:34:39.904 TRACE sink{name=output type=console}: vector::sinks::util::sink: submit successful. pending_acks=1
Apr 22 14:34:39.904 TRACE sink{name=output type=console}: vector::sinks::util::sink: Acking events. acking_num=1
{"message":"cat","parent":{"child":[]},"tags":[]}
```

I've also tested using `ndjson` encoding in the HTTP sink. In this case, lists simply become `[null]` et cetera. Removing `encoding.only_fields` fixes serialization altogether. 

### Expected Behavior

```
{"message":"cat","parent":{"child":["my-child"]},"tags":["my-tag"]}
```

### Example data

The following script should reproduce the issue:

```
#!/usr/bin/env bash
set -euo pipefail

cfg=faulty.yml

cat > "${cfg}" <<EOF
[sources.stdin]
type = "stdin"

[transforms.with_metadata]
type = "add_fields"
inputs = ["stdin"]
fields.tags = ["my-tag"]
fields.parent.child = ["my-child"]

[sinks.output]
type = "console"
inputs = ["with_metadata"]
encoding.codec = "json"
encoding.only_fields = ["message", "parent", "tags"]
EOF

echo "cat" | RUST_BACKTRACE=full vector -vvv -c "${cfg}"
```

### PR DESCRIPTION
**Pull request by Hoverbear**
Closes https://github.com/timberio/vector/issues/2407

Does not address https://github.com/timberio/vector/issues/2410. Future work is required there. Related to https://github.com/timberio/vector/issues/2414

### CODE REVIEW

File: src/sinks/util/encoding.rs [hunk 1]
```diff
@@ -156,7 +156,11 @@ pub trait EncodingConfiguration<E> {
                 Event::Log(log_event) => {
                     let to_remove = log_event
                         .keys()
-                        .filter(|f| !only_fields.contains(f))
+                        .filter(|f| {
+                            !only_fields
+                                .iter()
+                                .any(|only| f.starts_with(&(only.to_string() + "[")) || only == f)
```
Comment (Hoverbear) [inline src/sinks/util/encoding.rs:162]: Can you think of any ways to break this easily?

Comment (lukesteensen) [reply src/sinks/util/encoding.rs:163]: Can we use `PathIter` to get a `Vec<PathComponent>` and do prefix checking with that instead of strings?

Comment (Hoverbear) [reply src/sinks/util/encoding.rs:163]: Oh yesss that would work much better.

Comment (Hoverbear) [reply src/sinks/util/encoding.rs]: Ok that should be safer. :)

File: src/sinks/util/encoding.rs [hunk 2]
```diff
@@ -156,7 +159,23 @@ pub trait EncodingConfiguration<E> {
...
                         .keys()
-                        .filter(|f| !only_fields.contains(f))
+                        .filter(|field| {
+                            let field_path = PathIter::new(field).collect::<Vec<_>>();
+                            !only_fields.iter().any(|only| {
+                                if only == field {
+                                    true // fastpath
+                                } else {
+                                    // TODO: This is a hack for #2407, #2410 should fix this fully.
+                                    // This is *slow*.
```
Comment (binarylogic) [inline src/sinks/util/encoding.rs:169]: How slow?

File: src/sinks/util/encoding.rs [hunk 3]
```diff
@@ -156,7 +159,23 @@ pub trait EncodingConfiguration<E> {
                 Event::Log(log_event) => {
                     let to_remove = log_event
                         .keys()
-                        .filter(|f| !only_fields.contains(f))
+                        .filter(|field| {
+                            let field_path = PathIter::new(field).collect::<Vec<_>>();
```
Comment (binarylogic) [inline src/sinks/util/encoding.rs:163]: Can’t we skip this operation if only/except fields are blank?

File: src/sinks/util/encoding.rs [hunk 4]
```diff
@@ -156,7 +159,23 @@ pub trait EncodingConfiguration<E> {
...
+                            !only_fields.iter().any(|only| {
+                                if only == field {
+                                    true // fastpath
+                                } else {
+                                    // TODO: This is a hack for #2407, #2410 should fix this fully.
+                                    // This is *slow*.
+                                    let only_path = PathIter::new(only);
+                                    if field_path.starts_with(&only_path.collect::<Vec<_>>()[..]) {
+                                        true
+                                    } else {
```
Comment (binarylogic) [inline src/sinks/util/encoding.rs:173]: Don’t we need to parse the path for the except_fields loop below as well?

Comment (Hoverbear) [reply src/sinks/util/encoding.rs:173]: I checked just to be safe but we don't. I added the tests to this PR though.

image

**Comment by binarylogic** [comment]
Let's :shipit:

File: src/sinks/util/encoding.rs [hunk 5]
```diff
@@ -156,7 +159,23 @@ pub trait EncodingConfiguration<E> {
...
-                        .filter(|f| !only_fields.contains(f))
+                        .filter(|field| {
+                            let field_path = PathIter::new(field).collect::<Vec<_>>();
+                            !only_fields.iter().any(|only| {
+                                if only == field {
+                                    true // fastpath
+                                } else {
+                                    // TODO: This is a hack for #2407, #2410 should fix this fully.
+                                    // This is *slow*.
+                                    let only_path = PathIter::new(only);
```
Comment (lukesteensen) [inline src/sinks/util/encoding.rs:170]: If we're worried about perf here, could we do the iter and collect once at startup?

**Review by lukesteensen — APPROVED** [review]

File: src/sinks/util/encoding.rs (hunk 5 above)
Comment (binarylogic) [reply src/sinks/util/encoding.rs:170]: This seems worth it if true.

Comment (Hoverbear) [reply src/sinks/util/encoding.rs:170]: Yeah!!!! I was thinking about doing it right at startup time of the `EncodingConfiguration` itself, but thought that would be a bit invasive. Let me go ahead and just do it then.

**Review by lukesteensen — APPROVED** [review]
Looks good to me!

**Pull request merged**